olmo/checkpoint.py

```diff
@@ -611,49 +609,48 @@ class FullCheckpointer(Checkpointer):
     ) -> None:
         with self._temporary_wd(dir) as checkpoint_dir:
-            with FSDP.state_dict_type(
-                fsdp_model,
-                state_dict_type=StateDictType.FULL_STATE_DICT,
-                state_dict_config=FullStateDictConfig(rank0_only=True, offload_to_cpu=True),
-                optim_state_dict_config=FullOptimStateDictConfig(rank0_only=True, offload_to_cpu=True),
-            ):
-                # We'll write the model and optimizer state dicts individually to reduce (CPU) memory consumption.
-                # First the model state.
-                model_state_dict = fsdp_model.state_dict()
-                if get_global_rank() == 0:
-                    log.info("Saving model state...")
-                    save_state_dict(
-                        checkpoint_dir,
-                        "model.pt",
-                        model_state_dict,
-                        upload_to=upload_to,
-                        save_overwrite=self.cfg.save_overwrite,
-                        synchronize=False,
+            if isinstance(dist_model, FSDP):
+                with FSDP.state_dict_type(
+                    dist_model,
+                    state_dict_type=StateDictType.FULL_STATE_DICT,
+                    state_dict_config=FullStateDictConfig(rank0_only=True, offload_to_cpu=True),
+                    optim_state_dict_config=FullOptimStateDictConfig(rank0_only=True, offload_to_cpu=True),
+                ):
+                    # We'll write the model and optimizer state dicts individually to reduce (CPU) memory consumption.
+                    # First the model state.
+                    model_state_dict = dist_model.state_dict()
+                    self._write_model_dict(
+                        model_state_dict, checkpoint_dir, upload_to, save_overwrite=self.cfg.save_overwrite
                     )
-                del model_state_dict
-                barrier()
-
-                # Then the optimizer state.
-                optim_state_dict = FSDP.optim_state_dict(fsdp_model, optim)
-                if get_global_rank() == 0:
-                    log.info("Saving optim state...")
-                    save_state_dict(
-                        checkpoint_dir,
-                        "optim.pt",
-                        optim_state_dict,
-                        upload_to=upload_to,
-                        save_overwrite=self.cfg.save_overwrite,
-                        synchronize=False,
+
+                    # Then the optimizer state.
+                    optim_state_dict = FSDP.optim_state_dict(dist_model, optim)
+                    self._write_optim_dict(
+                        optim_state_dict, checkpoint_dir, upload_to, save_overwrite=self.cfg.save_overwrite
                     )
-                del optim_state_dict
-                barrier()
+            elif isinstance(dist_model, DDP):
+                # First, get the model state dict from DDP wrapped model
```

Comment:
Check whether we need to do something to prevent all the ranks from writing this file?

Reply:
_write_model_dict and _write_optim_dict have a check for this!

```
if get_global_rank() == 0:
```